### ISSUE TITLE
tensorflow.keras MirroredStrategy hangs without error message

### ISSUE THREAD

**Issue by WiestDaessle** [tensorflow/tensorflow#45178]
Unexpected hangs occurs, running adapted https://keras.io/guides/distributed_training/
**System information**
- Have I written custom code : https://keras.io/guides/distributed_training/ 
- OS Platform and Distribution : Windows 10 pro (build 19041) 
- CPU Intel: 2x Intel Xeon 6140 18 Cores (36 logical) 
- System Memory : 384 GB
- TensorFlow installed from (source or binary): pip installed
- TensorFlow version (use command below): v2.3.0-54-gfcc4b966f1 2.3.1
- Python version: 3.8.5
- CUDA/cuDNN version: CUDA 10.1 / cuDNN 7
- GPU model and memory: 4 x Tesla V100-SXM2-32GB 

**Describe the current behavior**
Training hangs without any message when starting the learning, when large dataset are used as inputs.

**Describe the expected behavior**

Training is expected to work in the same circumstances even with large dataset

**Standalone code to reproduce the issue**
Changing the get_dataset() to get a larger dataset from sample dataset following:

num_val_samples = 1000
and then adding the 2 lines:
    x_train = x_train[:-num_val_samples]
    y_train = y_train[:-num_val_samples]
    x_train = np.concatenate([x_train for i in range(50)])    
    y_train = np.concatenate([y_train for i in range(50)])    
    return (
        tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(batch_size),
        tf.data.Dataset.from_tensor_slices((x_val, y_val)).batch(batch_size),
        tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(batch_size),
    )

When the oversize is 46 it runs flawlessly, when setting to 47 or higher it just hangs with no error messages

The stragegy is defined with, to work on our Windows HPC : 
cross_tower_ops = tf.distribute. ReductionToOneDevice()
strategy = tf.distribute.MirroredStrategy(devices=['/gpu:0','/gpu:1','/gpu:2','gpu:3'],
                                            cross_device_ops=cross_tower_ops)

**Comment by WiestDaessle** [comment]
We also tried changing the data set with the strategy.experimental_distribute_dataset(val_dataset)

**Comment by nikitamaia** [comment]
Hi @WiestDaessle, to clarify, the training works okay when you run the tutorial? But it is hanging when you make these modifications? Is that correct?
Can you please provide the exact code you are running into issues with so I can reproduce it on my end. You can provide it in a colab notebook, or just paste it a comment, but the full code (properly indented and with import statements) greatly helps us to troubleshoot. Also please clarify what you mean by 'the oversize'.

Note that if you are using `MirroredStrategy` and `model.fit` then you should not need to use `strategy.experimental_distribute_dataset`. Distributing the dataset this way is only required when using a custom training loop.

**Comment by WiestDaessle** [comment]
Hi @nikitamaia, 
The sample is working after changing the strategy to point on our 4 GPUs using:
cross_tower_ops = tf.distribute.HierarchicalCopyAllReduce(num_packs=4)
strategy = tf.distribute.MirroredStrategy(devices=['/gpu:0','/gpu:1','/gpu:2','gpu:3'],
                                            cross_device_ops=cross_tower_ops)

**Comment by WiestDaessle** [comment]
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'  # or any {'0', '1', '2'}
import tensorflow as tf
from tensorflow import keras
import numpy as np

tf.get_logger().setLevel('DEBUG')

def get_compiled_model():
    # Make a simple 2-layer densely-connected neural network.
    inputs = keras.Input(shape=(784,))
    x = keras.layers.Dense(256, activation="relu")(inputs)
    x = keras.layers.Dense(256, activation="relu")(x)
    outputs = keras.layers.Dense(10)(x)
    model = keras.Model(inputs, outputs)
    model.compile(
        optimizer=keras.optimizers.Adam(),
        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
        metrics=[keras.metrics.SparseCategoricalAccuracy()],
    )
    return model


def get_dataset():
    batch_size = 10240
    num_val_samples = 10000

    # Return the MNIST dataset in the form of a `tf.data.Dataset`.
    (x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
    x_train = np.concatenate([x_train for i in range(50)])    
    y_train = np.concatenate([y_train for i in range(50)])

    # Preprocess the data (these are Numpy arrays)
    x_train = x_train.reshape(-1, 784).astype("float32") / 255
    x_test = x_test.reshape(-1, 784).astype("float32") / 255
    y_train = y_train.astype("float32")
    y_test = y_test.astype("float32")

    # Reserve num_val_samples samples for validation
    x_val = x_train[-num_val_samples:]
    y_val = y_train[-num_val_samples:]
    x_train = x_train[:-num_val_samples]
    y_train = y_train[:-num_val_samples]
    
    print(len(x_train), len(y_train),len(x_val),len(y_val))    
    return (
        tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(batch_size),
        tf.data.Dataset.from_tensor_slices((x_val, y_val)).batch(batch_size),
        tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(batch_size),
    )


# Create a MirroredStrategy.
#strategy = tf.distribute.MirroredStrategy()
# cross_tower_ops = tf.distribute.ReductionToOneDevice()
cross_tower_ops = tf.distribute.HierarchicalCopyAllReduce(num_packs=4)
# Reduce to CPU
#cross_tower_ops = tf.distribute. ReductionToOneDevice(reduce_to_device="/device:CPU:0")
strategy = tf.distribute.MirroredStrategy(devices=['/gpu:0','/gpu:1','/gpu:2','gpu:3'],
                                            cross_device_ops=cross_tower_ops)
print("Number of devices: {}".format(strategy.num_replicas_in_sync))

# Open a strategy scope.
with strategy.scope():
    # Everything that creates variables should be under the strategy scope.
    # In general this is only model construction & `compile()`.
    model = get_compiled_model()
    # model.summary()

# Train the model on all available devices.
train_dataset, val_dataset, test_dataset = get_dataset()
    # train_dataset, val_dataset, test_dataset = (strategy.experimental_distribute_dataset(train_dataset),
    #         strategy.experimental_distribute_dataset(val_dataset), 
    #         strategy.experimental_distribute_dataset(test_dataset))
    # model.summary()
model.fit(train_dataset, epochs=200, validation_data=val_dataset)

# Test the model on all available devices.
model.evaluate(test_dataset)

**Comment by nikitamaia** [comment]
@WiestDaessle please provide the code with the proper indentation and formatting to help with the debugging process. Thanks!

**Comment by WiestDaessle** [comment]
fucking markdown not taking the indent into account,
should be ok now

**Comment by nikitamaia** [comment]
I tried to run the code (with 2 V100s) and I saw that it was hanging on `train_dataset, val_dataset, test_dataset = get_dataset()` making me suspect that this is not an issue with the distribution strategy. Can you just try and execute your `get_datatset()` function without a strategy and see if it still hangs? I seem to be hitting a memory issue with `tf.data.Dataset.from_tensor_slices((x_train, y_train)).batch(batch_size)`
